apps/devhub/tests/test_utils.py

```diff
@@ -0,0 +1,373 @@
+
+        assert not self.validate_upload.called
+        self.validate_file.assert_has_calls([call([self.file_1_1.pk]),
+                                             call([self.file.pk])])
+
+        self.save_file.assert_called_once_with([self.file_1_1.pk])
+
+    def test_find_future_fileupload_version(self):
+        """Test that a future version will not be matched."""
+
+        self.version.update(version='1.2')
+
+        va = utils.ValidationAnnotator(self.file_upload)
+        assert va.find_previous_version(self.xpi_version) is None
+
+        assert not self.validate_file.called
+        self.validate_upload.assert_called_once_with(
+            [self.file_upload.pk, None])
+
+        self.save_upload.assert_called_once_with([self.file_upload.pk])
+
+    def test_find_future_file(self):
+        """Test that a future version will not be matched."""
+
+        self.version.update(version='1.2')
+
+        va = utils.ValidationAnnotator(self.file_1_1)
+        assert va.find_previous_version(self.xpi_version) is None
+
+        assert not self.validate_upload.called
+        self.validate_file.assert_called_once_with([self.file_1_1.pk])
+
+        self.save_file.assert_called_once_with([self.file_1_1.pk])
+
+
+class TestValidationAnnotatorBeta(TestValidationAnnotatorBase):
+
+    def setUp(self):
+        super(TestValidationAnnotatorBeta, self).setUp()
+
+        self.addon.update(is_listed=True)
+
+        self.xpi_version = '1.1b1'
+
+        parse_addon = self.patch('devhub.utils.parse_addon')
+        parse_addon.return_value = {'version': self.xpi_version,
+                                    'guid': self.addon.guid}
+
+    def test_match_beta_to_release(self):
+        """Test that a beta submission is matched to the latest approved
+        release version."""
+
+        va = utils.ValidationAnnotator(self.file_upload)
+        assert va.find_previous_version(self.xpi_version) == self.file
+
+        self.validate_file.assert_called_once_with([self.file.pk])
+
+    def test_match_beta_to_signed_beta(self):
+        """Test that a beta submission is matched to a prior signed beta
+        version."""
+
+        self.file_1_1.update(status=amo.STATUS_BETA,
+                             is_signed=True)
+        self.version_1_1.update(version='1.1b0')
+
+        va = utils.ValidationAnnotator(self.file_upload)
+        assert va.find_previous_version(self.xpi_version) == self.file_1_1
+
+        self.validate_file.assert_called_once_with([self.file_1_1.pk])
+
+    def test_match_beta_to_unsigned_beta(self):
+        """Test that a beta submission is not matched to a prior unsigned beta
+        version."""
+
+        self.file_1_1.update(status=amo.STATUS_BETA)
+        self.version_1_1.update(version='1.1b0')
+
+        va = utils.ValidationAnnotator(self.file_upload)
+        assert va.find_previous_version(self.xpi_version) == self.file
+
+        self.validate_file.assert_called_once_with([self.file.pk])
+
```

Comment:
At the moment, no, since none of the annotations that we do are currently relevant to those add-ons. At some point, I do want to update this code to find the correct version to compare against in the review tools (https://bugzil.la/666976) and to flag messages that are unchanged since that version (there's another bug for this, which I can't find right now). So those tests should be added when we implement that behavior.